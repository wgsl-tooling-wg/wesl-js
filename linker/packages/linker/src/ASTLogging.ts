import { AbstractElem2, ModuleElem } from "./AbstractElems2.ts";
import { treeToString } from "./ImportTree.ts";
import { LineWrapper } from "./LineWrapper.ts";

export function astTree(elem: AbstractElem2, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astTree(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

function addElemFields(elem: AbstractElem2, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addNameFields(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addImport(elem, str) ||
    addIdentFields(elem, str);
}

function addVarishFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (kind === "var" || kind === "const" || kind === "override") {
    str.add(" " + elem.name.ident.originalName);
    if (elem.typeRef) {
      str.add(":" + elem.typeRef.ident.originalName);
    }
    return true;
  }
}

function addTextFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "text") {
    const { src, start, end } = elem;
    str.add(` '${src.slice(start, end)}'`);
    return true;
  }
}

function addIdentFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "ident") {
    const { ident } = elem;
    const prefix = ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + elem.ident.originalName);
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRef.ident.originalName);
    return true;
  }
}

function addStructFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addImport(elem: AbstractElem2, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + treeToString(elem.imports));
    return true;
  }
}

function addNameFields(
  elem: AbstractElem2,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem2, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType } = elem;
    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(p => p.name.ident.originalName + ": " + p.typeRef.ident.originalName)
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    if (returnType) {
      str.add(" -> " + returnType.ident.originalName);
    }

    return true;
  }
}
