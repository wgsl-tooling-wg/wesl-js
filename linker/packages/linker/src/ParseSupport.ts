import {
  any,
  anyNot,
  anyThrough,
  disablePreParse,
  ExtendedResult,
  kind,
  makeEolf,
  or,
  Parser,
  repeat,
  req,
  resultLog,
  seq,
  setTraceName,
  tokens,
  tracing,
} from "mini-parse";
import { argsTokens, lineCommentTokens, mainTokens } from "./WESLTokens.ts";

/* Basic parsing functions for comment handling, eol, etc. */

export const word = or(
  kind(mainTokens.ident),
  kind(mainTokens.textureStorage),
);

export const unknown = any().map(r => {
  const { kind, text } = r.value;
  const deepName = r.ctx._debugNames.join(" > ");

  resultLog(r, `??? ${kind}: '${text}'  ${deepName}`);
  // throw new Error("Fail fast");
});

export const blockComment: Parser<any> = seq(
  "/*",
  repeat(or(() => blockComment, anyNot("*/"))),
  req("*/"),
);

export const eolf: Parser<any> = disablePreParse(
  makeEolf(argsTokens, argsTokens.ws),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

export const comment = or(() => lineComment, blockComment).trace({
  hide: true,
});

