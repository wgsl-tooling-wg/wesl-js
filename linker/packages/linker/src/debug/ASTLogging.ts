import { AbstractElem, ModuleElem } from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

export function astTree(elem: AbstractElem, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astTree(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

export function elemToString(elem: AbstractElem): string {
  const { kind } = elem as ModuleElem;
  const str = new LineWrapper();
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }
  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addNameFields(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addImport(elem, str) ||
    addRefIdent(elem, str) ||
    addDeclIdent(elem, str);
}

function addVarishFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (
    kind === "var" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    str.add(" " + elem.name.ident.originalName);
    if (elem.typeRef) {
      str.add(":" + elem.typeRef.ident.originalName);
    }
    return true;
  }
}

function addTextFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
    return true;
  }
}

function addRefIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "ref") {
    str.add(" " + elem.ident.originalName);
    return true;
  }
}

function addDeclIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "decl") {
    str.add(" %" + elem.ident.originalName);
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRef.ident.originalName);
    return true;
  }
}

function addStructFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addImport(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + importToString(elem.imports));
    return true;
  }
}

function addNameFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType } = elem;
    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(p => p.name.ident.originalName + ": " + p.typeRef.ident.originalName)
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    if (returnType) {
      str.add(" -> " + returnType.ident.originalName);
    }

    return true;
  }
}
