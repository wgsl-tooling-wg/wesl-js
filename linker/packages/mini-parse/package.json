{
  "name": "mini-parse",
  "version": "0.5.0-d1",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "run-s build:main build:util build:minified build:brotli",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:minified": "vite -c minified.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs && ls -l dist/minified.cjs.br",
    "format": "prettier . --write",
    "lint": "eslint src",
    "organize": "organize-imports-cli tsconfig.json",
    "prepublishOnly": "run-s build",
    "test": "vitest",
    "test:once": "vitest run"
  },
  "type": "module",
  "repository": "github:wgsl-tooling-wg/wgsl-linker",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./minified": {
      "types": "./dist/index.d.ts",
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^22.5.0",
    "@typescript-eslint/eslint-plugin": "^8.2.0",
    "@typescript-eslint/parser": "^8.2.0",
    "@vitest/ui": "^2.0.5",
    "berry-pretty": "^0.0.5",
    "esbuild": "^0.23.1",
    "eslint": "^9.9.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.31.6",
    "typescript": "^5.5.4",
    "vite": "^5.4.2",
    "vite-plugin-dts": "^4.0.3",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.0.5"
  }
}
