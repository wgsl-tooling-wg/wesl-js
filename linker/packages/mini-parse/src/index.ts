export * from "./MatchingLexer.js";
export * from "./Parser.js";
export * from "./ParserCollect.js";
export * from "./ParserCombinator.js";
export * from "./ParserLogging.js";
export * from "./ParserToString.js";
export * from "./ParserTracing.js";
export * from "./Span.js";
export * from "./SrcMap.js";
export * from "./SrcMapBuilder.js";
export * from "./TokenMatcher.js";
export * from "./WrappedLog.js";
