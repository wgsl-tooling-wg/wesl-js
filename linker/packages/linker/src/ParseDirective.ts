import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "mini-parse";
import { gleamImport } from "./GleamImport.js";
import { argsTokens, lineCommentTokens, mainTokens } from "./MatchWgslD.js";
import { eolf } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq("from", or(fromWord, seq('"', fromWord, '"')));

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf));

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, gleamImport)),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    skipToEol,
    lineComment,
    directive,
  });
}
