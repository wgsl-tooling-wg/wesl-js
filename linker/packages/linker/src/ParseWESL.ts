import { AppState, matchingLexer, ParserInit, SrcMap } from "mini-parse";
import { AbstractElem } from "./AbstractElems.ts";
import { AbstractElem2, ModuleElem } from "./AbstractElems2.ts";
import { mainTokens } from "./MatchWgslD.ts";
import { emptyScope, resetScopeIds, Scope, SrcModule } from "./Scope.ts";
import { OpenElem } from "./WESLCollect.ts";
import { weslRoot } from "./WESLGrammar.ts";

/** result of a parse */
export interface WeslAST {
  elems: AbstractElem[]; // legacy
  rootModule: ModuleElem;
  scope: Scope;
}

/** stable and unstable state used during parsing */
export interface WeslParseState extends AppState<WeslParseContext> {
  context: WeslParseContext;
  stable: StableState;
}

/** stable values used or accumulated during parsing */
export interface StableState {
  // parameters for evaluating conditions while parsing this module
  conditions: Record<string, any>;

  // legacy elems succesfully parsed in this module
  elems: AbstractElem[];

  // elems succesfully parsed in this module
  rootModule?: ModuleElem;

  // root scope for this module
  rootScope: Scope;
}

/** unstable values used during parse collection */
export interface WeslParseContext {
  scope: Scope; // current scope (points somewhere in rootScope)
  openElems: OpenElem[]; // elems that are collecting their contents
}

export function parseSrcModule(
  srcModule: SrcModule,
  maxParseCount: number | undefined = undefined,
): WeslAST {
  const lexer = matchingLexer(srcModule.src, mainTokens);

  const appState = blankWeslParseState();

  const init: ParserInit = { lexer, appState, maxParseCount };
  const parseResult = weslRoot.parse(init);
  if (parseResult === null) {
    throw new Error("parseWESL failed");
  }

  const { rootModule, elems, rootScope } = appState.stable;
  return { rootModule: rootModule!, scope: rootScope, elems };
}

// TODO make wrapper on srcModule
export function parseWESL(
  src: string,
  srcMap?: SrcMap,
  conditions: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = weslRoot,
): WeslAST {
  // TODO allow returning undefined for failure, or throw?

  resetScopeIds();
  const lexer = matchingLexer(src, mainTokens);

  const appState = blankWeslParseState();

  const init: ParserInit = { lexer, appState: appState, srcMap, maxParseCount };
  const parseResult = grammar.parse(init);
  if (parseResult === null) {
    throw new Error("parseWESL failed");
  }

  const { rootModule, elems, rootScope } = appState.stable;
  return { rootModule: rootModule!, scope: rootScope, elems };
}

export function blankWeslParseState(): WeslParseState {
  const rootScope = emptyScope("module");
  return {
    context: { scope: rootScope, openElems: [] },
    stable: { conditions: {}, elems: [], rootScope },
  };
}
