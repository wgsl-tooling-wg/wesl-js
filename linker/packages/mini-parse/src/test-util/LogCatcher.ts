import { withLogger, withLoggerAsync } from "mini-parse";

export interface LogCatcher {
  /** tests can use this to replace console.log with a log capturing function */
  log: (...params: any[]) => void;
  logged: () => string;
}

export function logCatch(): LogCatcher {
  const lines: string[] = [];
  function log(...params: any[]): void {
    lines.push(params.join(" "));
  }
  function logged(): string {
    return lines.join("\n");
  }
  return { log, logged };
}

export function withLogSpy(fn: () => void): string {
  const catcher = logCatch();
  withLogger(c => catcher.log(c), fn);
  return catcher.logged();
}

export async function withLogSpyAsync(
  fn: () => Promise<void>,
): Promise<string> {
  const catcher = logCatch();
  await withLoggerAsync(c => catcher.log(c), fn);
  return catcher.logged();
}
