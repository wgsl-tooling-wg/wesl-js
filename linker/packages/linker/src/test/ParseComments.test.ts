import { expectNoLog } from "mini-parse/test-util";

import { expect, test } from "vitest";
import { lineComment } from "../ParseDirective.js";
import { blockComment } from "../ParseSupport.js";
import { parseWESL } from "../ParseWESL.js";
import { testAppParse } from "./TestUtil.js";
import { astTree } from "../debug/ASTLogging.js";

test("lineComment parse // foo bar", () => {
  const src = "// foo bar";
  const { position } = testAppParse(lineComment, src);
  expect(position).toBe(src.length);
});

test("lineComment parse // foo bar \\n", () => {
  const src = "// foo bar\n";
  const { position } = testAppParse(lineComment, src);
  expect(position).toBe(src.length);
});

test("blockComment parses /* comment */", () => {
  const src = "/* comment */";
  expectNoLog(() => {
    const { parsed } = testAppParse(blockComment, src);
    expect(parsed?.value).toMatchInlineSnapshot(`
      [
        "/*",
        [
          {
            "kind": "ident",
            "text": "comment",
          },
        ],
        "*/",
      ]
    `);
  });
});

test("skipBlockComment parses nested comment", () => {
  const src = "/** comment1 /* comment2 */ */";
  expectNoLog(() => testAppParse(blockComment, src));
});

test("parse fn with line comment", () => {
  const src = `
    fn binaryOp() { // binOpImpl
    }`;
  const parsed = parseWESL(src);
  expect(astTree(parsed.moduleElem)).toMatchInlineSnapshot(`
    "module
      text '
        '
      fn binaryOp()
        text 'fn '
        decl %binaryOp
        text '() { // binOpImpl
        }'"
  `);
});

test("parse empty line comment", () => {
  const src = `
  var workgroupThreads= 4;                          // 
  `;
  expectNoLog(() => parseWESL(src));
});

test("parse line comment with #replace", () => {
  const src = ` 
  const workgroupThreads= 4;                          // #replace 4=workgroupThreads
  `;
  expectNoLog(() => parseWESL(src));
});
