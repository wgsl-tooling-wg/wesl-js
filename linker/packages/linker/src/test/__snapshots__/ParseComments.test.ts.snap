// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`blockComment parses /* comment */ 1`] = `
{
  "tags": {},
  "value": [
    "/*",
    [
      {
        "kind": "ident",
        "text": "comment",
      },
    ],
    "*/",
  ],
}
`;
